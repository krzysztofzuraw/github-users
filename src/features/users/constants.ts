export const GET_USERS_REQUEST = "users/GET_USERS_REQUEST";
export const GET_USERS_SUCCESS = "users/GET_USERS_SUCCESS";
export const GET_USERS_ERROR = "users/GET_USERS_ERROR";
