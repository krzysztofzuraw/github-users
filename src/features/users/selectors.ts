import { createSelector } from "reselect";

import Types from "Types";

const getUserState = (state: Types.RootState) => state.users;

const getUserById = (state: Types.RootState, userId: string) => state.users.byId[userId];

export const getUsers = createSelector(getUserState, state =>
  state.allIds.map(id => state.byId[id]).slice(0, 3)
);

export const usersFetching = createSelector(getUserState, state => state.isFetching);
