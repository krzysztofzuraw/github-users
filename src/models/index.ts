export * from "@src/models/user";
export * from "@src/models/locales";
