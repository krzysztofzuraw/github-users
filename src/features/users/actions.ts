import { action } from "typesafe-actions";

import { IUser } from "@src/models";
import { GET_USERS_ERROR, GET_USERS_REQUEST, GET_USERS_SUCCESS } from "./constants";

export const getUsersRequest = () => action(GET_USERS_REQUEST);
export const getUsersSuccess = (users: IUser[]) => action(GET_USERS_SUCCESS, users);
export const getUsersError = (message: string) => action(GET_USERS_ERROR, message);
