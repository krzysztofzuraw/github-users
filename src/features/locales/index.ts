import * as localesActions from "@src/features/locales/actions";
import * as localesConstants from "@src/features/locales/constants";
import localesEpic from "@src/features/locales/epics";
import localesReducer, { LocalesActions, LocalesState } from "@src/features/locales/reducers";
import * as localesSelectors from "@src/features/locales/selectors";

export {
  localesActions,
  localesConstants,
  localesReducer,
  LocalesActions,
  LocalesState,
  localesEpic,
  localesSelectors,
};
