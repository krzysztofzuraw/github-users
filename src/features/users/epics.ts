import { combineEpics, Epic } from "redux-observable";
import { Observable } from "rxjs/Observable";
import Types from "Types";

import { usersActions, usersConstants } from "../users";

const fetchUsers: Epic<Types.RootAction, Types.RootState, Types.Services> = (
  action$,
  store,
  { githubService, loggerService }
) =>
  action$.ofType(usersConstants.GET_USERS_REQUEST).switchMap(() =>
    Observable.from(githubService.getUsers())
      .map(response => usersActions.getUsersSuccess(response))
      .catch(error => Observable.of(usersActions.getUsersError(error)))
  );

export default combineEpics(fetchUsers);
