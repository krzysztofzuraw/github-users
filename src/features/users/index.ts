import * as usersActions from "./actions";
import * as usersConstants from "./constants";
import userEpic from "./epics";
import userReducer, { UsersActions, UsersState } from "./reducers";
import * as usersSelectors from "./selectors";

export {
  userEpic,
  userReducer,
  usersActions,
  usersConstants,
  UsersState,
  UsersActions,
  usersSelectors,
};
