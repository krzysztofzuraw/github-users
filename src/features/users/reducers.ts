import { combineReducers } from "redux";
import Types from "Types";
import { ActionType } from "typesafe-actions";

import { usersActions, usersConstants } from "@src/features/users";
import { IUser } from "@src/models";

export type UsersState = Readonly<{
  allIds: string[];
  byId: { [id: string]: IUser };
  isFetching: boolean;
}>;

export type UsersActions = ActionType<typeof usersActions>;

const allIds: Types.Reducer<UsersState["allIds"]> = (state = [], action) => {
  switch (action.type) {
    case usersConstants.GET_USERS_SUCCESS:
      return action.payload.map(user => user.id);
    default:
      return state;
  }
};

const byId: Types.Reducer<UsersState["byId"]> = (state = {}, action) => {
  switch (action.type) {
    case usersConstants.GET_USERS_SUCCESS:
      return action.payload.reduce<UsersState["byId"]>(
        (result, item) => ((result[item.id] = item), result),
        {}
      );
    default:
      return state;
  }
};

const isFetching: Types.Reducer<UsersState["isFetching"]> = (state = false, action) => {
  switch (action.type) {
    case usersConstants.GET_USERS_REQUEST:
      return true;
    case usersConstants.GET_USERS_SUCCESS:
    case usersConstants.GET_USERS_ERROR:
      return false;
    default:
      return state;
  }
};

const userReducer = combineReducers<UsersState>({ allIds, byId, isFetching });

export default userReducer;
