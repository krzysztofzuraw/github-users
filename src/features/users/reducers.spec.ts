import { userReducer, usersActions } from "@src/features/users";
import { User } from "@src/models";

const initialState = { allIds: [], byId: {}, isFetching: false };

describe("Users reducer", () => {
  it("should match snapshot for inital state", () => {
    expect(initialState).toMatchSnapshot();
  });

  describe("allIds, byId", () => {
    const user = new User("log", "http://", "http://", "name", "http://", 1, "");
    const action = usersActions.getUsersSuccess([user]);

    it("should return all user ids", () => {
      const state = userReducer(initialState, action);
      expect(state.allIds).toEqual([user.id]);
    });

    it("should return user by their id", () => {
      const state = userReducer(initialState, action);
      expect(state.byId).toEqual({ [user.id]: user });
    });
  });

  describe("isFetching", () => {
    it("should return true when request is in progress", () => {
      const state = userReducer(initialState, usersActions.getUsersRequest());
      expect(state.isFetching).toBeTruthy();
    });

    it("should return false when request ends with success", () => {
      const user = new User("log", "http://", "http://", "name", "http://", 1, "");
      const state = userReducer(initialState, usersActions.getUsersSuccess([user]));
      expect(state.isFetching).toBeFalsy();
    });

    it("should return false when request ends with error", () => {
      const state = userReducer(initialState, usersActions.getUsersError("API Error"));
      expect(state.isFetching).toBeFalsy();
    });
  });
});
