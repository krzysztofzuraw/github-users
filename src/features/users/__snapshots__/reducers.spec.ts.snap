// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users reducer should match snapshot for inital state 1`] = `
Object {
  "allIds": Array [],
  "byId": Object {},
  "isFetching": false,
}
`;
